main.go

```diff
@@ -1024,48 +1031,115 @@ type funcFullName = string
 // parameters, so we can avoid obfuscating types used with them.
 //
-// For now, this table is manually maintained, until we automatically detect and
-// propagate the uses of reflect in std and third party APIs.
-// If we end up maintaining this table for a long time, we should probably
-// improve the process via either code generation or sanity checks.
-//
-// TODO: record which parameters get used with reflection, as it's often not all
-// of them.
-//
 // TODO: we're not including fmt.Printf, as it would have many false positives,
 // unless we were smart enough to detect which arguments get used as %#v or %T.
-//
-// TODO: this should also support third party APIs; see
-// https://github.com/burrowers/garble/issues/162
-var knownReflectAPIs = map[funcFullName]bool{
-	"reflect.TypeOf":  true,
-	"reflect.ValueOf": true,
-
-	"encoding/json.Marshal":           true,
-	"encoding/json.MarshalIndent":     true,
-	"encoding/json.Unmarshal":         true,
-	"(*encoding/json.Decoder).Decode": true,
-	"(*encoding/json.Encoder).Encode": true,
-
-	"encoding/gob.Register":          true,
-	"encoding/gob.RegisterName":      true,
-	"(*encoding/gob.Decoder).Decode": true,
-	"(*encoding/gob.Encoder).Encode": true,
-
-	"encoding/xml.Marshal":                  true,
-	"encoding/xml.MarshalIndent":            true,
-	"encoding/xml.Unmarshal":                true,
-	"(*encoding/xml.Decoder).Decode":        true,
-	"(*encoding/xml.Encoder).Encode":        true,
-	"(*encoding/xml.Decoder).DecodeElement": true,
-	"(*encoding/xml.Encoder).EncodeElement": true,
-
-	"(*text/template.Template).Execute": true,
-
-	"(*html/template.Template).Execute": true,
-
-	"net/rpc.Register":     true,
-	"net/rpc.RegisterName": true,
+var knownReflectAPIs = map[funcFullName][]int{
+	"reflect.TypeOf":  {0},
+	"reflect.ValueOf": {0},
+}
+
+func loadKnownReflectAPIs() error {
+	dir := filepath.Join(os.TempDir(), "garble")
+	filename := filepath.Join(dir, "reflect.json")
+
+	data, err := os.ReadFile(filename)
+	if err != nil || len(data) == 0 {
+		return nil
+	}
+
+	return json.Unmarshal(data, &knownReflectAPIs)
+}
+
+func saveKnownReflectAPIs() error {
+	dir := filepath.Join(os.TempDir(), "garble")
+	filename := filepath.Join(dir, "reflect.json")
+
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return err
+	}
+
+	data, err := json.Marshal(knownReflectAPIs)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(filename, data, 0666)
+}
+
+func (tf *transformer) findReflectFunctions(files []*ast.File) {
+	visitReflect := func(node ast.Node) bool {
+		funcDecl, ok := node.(*ast.FuncDecl)
+		if !ok {
+			return true
+		}
+
+		funcObj := tf.info.ObjectOf(funcDecl.Name).(*types.Func)
+
+		var paramNames []string
+		for _, param := range funcDecl.Type.Params.List {
+			for _, name := range param.Names {
+				paramNames = append(paramNames, name.Name)
+			}
+		}
+
+		ast.Inspect(funcDecl, func(node ast.Node) bool {
```

Comment:
you do an ast.Inspect inside another ast.Inspect; that's almost always unnecessary and somewhat wasteful, unless for some reason you jump from walking one file to another.

Reply:
Right will fix

Reply:
Hmm, this isn't fixed, right?

Reply:
Is fixed because the code calling this doesn't use ast.Inspect anymore